spring-xd-shell/src/main/java/org/springframework/xd/shell/command/ModuleCommands.java

```diff
@@ -141,10 +157,31 @@ public String uploadModule(
 			@CliOption(mandatory = true, key = {"", "file"}, help = "path to the module archive") File file,
 			@CliOption(key = "force", help = "force update if module already exists (only if not in use)", specifiedDefaultValue = "true", unspecifiedDefaultValue = "false") boolean force
 	) throws IOException {
-		Resource resource = new FileSystemResource(file);
-		ModuleDefinitionResource composedModule = moduleOperations().uploadModule(name, type, resource, force);
-		return String.format(("Successfully uploaded module '%s:%s'"), composedModule.getType(),
-				composedModule.getName());
+		Resource resource;
+		if (file.isDirectory()) {
+			PipedOutputStream pos = new PipedOutputStream();
+			resource = new InputStreamResource(new PipedInputStream(pos)) {
+				@Override
+				public String getFilename() {
+					return "some.jar"; // To trick media-type recognition
+				}
+
+				@Override
+				public long contentLength() throws IOException {
+					return -1; // and this is to prevent multiple stream reads, as
+					// this is now effectively a subclass of ISResource
+				}
+			};
+			final JarOutputStream jos = new JarOutputStream(pos);
+			new Thread(new DirectoryJarBuilder(file, jos)).start();
```

Comment:
Since this is async, if this fails, we get a success message back from the command anyway. Does it need to run in a thread?